Updated json output to make it more easily parsable

Devaniti opened the pull request:
- Added option to output info from all adapters
- Implemented querying of D3D12_FEATURE_DATA_COMMAND_QUEUE_PRIORITY
- Refactored header structure
- Changed OS version output
  - Removed unused or irrelevant parameters (service pack related parameters, product type related parameters)
  - Changed formatting to output version in human readable format
- Removed reporting of DXGI_FEATURE_PRESENT_ALLOW_TEARING
	- It's an OS capability. It was introduced before Agility SDK support, so D3d12info won't be useful on systems that has it set to false anyway.

Example output before PR:
[2.2.0.txt](https://github.com/sawickiap/D3d12info/files/13774002/2.2.0.txt)
[2.2.0.json](https://github.com/sawickiap/D3d12info/files/13774005/2.2.0.json)
Example output after PR:
[PR.txt](https://github.com/sawickiap/D3d12info/files/13774001/PR.txt)
[PR.json](https://github.com/sawickiap/D3d12info/files/13774004/PR.json)
[PR_all.txt](https://github.com/sawickiap/D3d12info/files/13773999/PR_all.txt)
[PR_all.json](https://github.com/sawickiap/D3d12info/files/13774003/PR_all.json)

sawickiap commented:
Thank you very much for this extensive pull request! I like most of the changes you proposed. I would like to discuss few topics.

## Things I like

Change "Built" to "BuildDate".

Change of OSVERSIONINFOEX OS Info / Windows version - I trust you this is OK and I hope this really is the main Windows version that gets updates after Windows Update.

Removing `DXGI_FEATURE_PRESENT_ALLOW_TEARING` - OK, I trust you if you say it doesn't make sense to query.

`D3D12_FEATURE_DATA_COMMAND_QUEUE_PRIORITY` - thank you for adding it! It was missing.

Change in JSON format to print array `[]` of adapters with "AdapterIndex" inside - great idea!

`CmdLineParser::m_EncounteredOpts`: Great idea!

## Things I like to discuss

Printing all adapters: I like the idea overall, but I don't like printing "Microsoft Basic Render Driver" by default. How about following convention:
- By default, we print all non-SOFTWARE adapters.
- With parameter `--AllAdapters`, we print all, including SOFTWARE.

By the way, parameters starting with single `-` are intended to be single letter, so `-all` breaks this convention.

Change in UMDVersion: The way you print it breaks the convention used everywhere else that we have structs with parameters inside. It gets printed on the top level instead of inside a structure.

Printing UMDVersion like "31.0.15.4633" also in JSON: I would like to continue a convention that data in JSON is as raw as possible and only formatted into some human-friendly form during further processing, just like I print sizes in JSON as bytes, not converted to "X MB".

`throw std::runtime_error("Could not detect Windows version");` - I would like to continue a convention that if fetching some information fails, we simply skip printing that information instead of failing entire program.

Devaniti commented:
One clarification on OSVERSIONINFOEX OS Info / Windows version
That format shows all data from OSVERSIONINFOEX that is useful in modern changes of Windows
But OSVERSIONINFOEX alone is not enough. There's one additional number after build number in Windows version that is not in OSVERSIONINFOEX. There are also updates that doesn't contribute to version number, though I'm not sure whether those can change d3d/dxgi behavior.

And here's what I changed in last revision:
* Returned back `DXGI_FEATURE_PRESENT_ALLOW_TEARING`
	* I still expect it to be true on all systems that support Agility SDK, but there's just no other tool to check value of that
	* I moved it from Adapter to System Info section
* Changed command line parameters
	* Current list:
		* `--AllNonSoftware     Print details of all (except WARP and Software) adapters (default behavior).`
		* `--AllAdapters        Print details of all (except WARP) adapters.`
	* Having `--AllNonSoftware` makes it easy to document what's default behaviour
	* Removed `--all`, otherwise it would have been ambigious
		* `-all` parameter was a typo, it supposed to be `--all`
		* D3d12info does not support single dash parameters of more than 1 character so couldn't have `-all` in the first place
* Changed windows version detection failure to not prevent other info from fetching
	* I found a bug in the process, that function had `PrintStructBegin` and early out without respective `PrintStructEnd`
		* Fixed it by making ScopedStructRegion struct that automatically calls `PrintStructBegin`/`PrintStructEnd` in constructor/destructor to prevent this bug in the future. 
			* Replaced all `PrintStructBegin`/`PrintStructEnd` calls with `ScopedStructRegion`

Devaniti commented:
Sample output on new revision
[PR-AllAdapters.txt](https://github.com/sawickiap/D3d12info/files/13798317/PR-AllAdapters.txt)
[PR-AllAdapters.json](https://github.com/sawickiap/D3d12info/files/13798318/PR-AllAdapters.json)
[PR.json](https://github.com/sawickiap/D3d12info/files/13798319/PR.json)
[PR.txt](https://github.com/sawickiap/D3d12info/files/13798320/PR.txt)

Devaniti commented:
Oh, and UMDVersion is also changed
I've reverted all changes related to it except names
It's now inside CheckInterfaceSupport structure and is called UMDVersion in both json and non-json codepaths

sawickiap commented:
I am sorry for delayed response. New Year is not a good time for coding 😉🍺

Thank you very much for high quality code. I merged your pull request. 

Great idea with the `ScopedStructRegion`!

I made one change - header prints `D3D12_SDK_VERSION` or `D3D12_PREVIEW_SDK_VERSION`. I know it will complicate your parsing, but in "preview" version I don't want to print it as `D3D12_SDK_VERSION` because it is simply not correct.

I also noticed that after changing logic to print multiple adapters by default, `--WARP` parameter stopped working. I fixed it.
